Sources/NIOPosix/VsockAddress.swift

```diff
@@ -0,0 +1,290 @@
+    /// - seealso: `LocalVsockContextID`
+    public static let localVsockContextID = Types.LocalVsockContextID()
+}
+
+extension ChannelOptions.Types {
+    /// This get-only option is used on channels backed by vsock sockets to get the local VSOCK context ID.
+    public struct LocalVsockContextID: ChannelOption, Sendable {
+        public typealias Value = VsockAddress.ContextID
+        public init() {}
+    }
+}
+
+// MARK: - Public API that might throw runtime error if not implemented on the platform.
+
+extension NIOBSDSocket.AddressFamily {
+    /// Address for vsock.
+    public static var vsock: NIOBSDSocket.AddressFamily {
+#if canImport(Darwin) || os(Linux)
+        NIOBSDSocket.AddressFamily(rawValue: AF_VSOCK)
+#else
+        fatalError(vsockUnimplemented)
+#endif
+    }
+}
+
+extension NIOBSDSocket.ProtocolFamily {
+    /// Address for vsock.
+    public static var vsock: NIOBSDSocket.ProtocolFamily {
+#if canImport(Darwin) || os(Linux)
+        NIOBSDSocket.ProtocolFamily(rawValue: PF_VSOCK)
+#else
+        fatalError(vsockUnimplemented)
+#endif
+    }
+}
+
+extension VsockAddress {
+    public func withSockAddr<T>(_ body: (UnsafePointer<sockaddr>, Int) throws -> T) rethrows -> T {
+#if canImport(Darwin) || os(Linux)
+        return try self.address.withSockAddr({ try body($0, $1) })
+#else
+        fatalError(vsockUnimplemented)
+#endif
+    }
+}
+
+// MARK: - Internal functions that are only available on supported platforms.
+
+#if canImport(Darwin) || os(Linux)
+extension VsockAddress.ContextID {
+    /// Get the context ID of the local machine.
+    ///
+    /// - Parameters:
+    ///   - socketFD: the file descriptor for the open socket.
+    ///
+    /// This function wraps the `IOCTL_VM_SOCKETS_GET_LOCAL_CID` `ioctl()` request.
+    ///
+    /// To provide a consistent API on Linux and Darwin, this API takes a socket parameter, which is unused on Linux:
+    ///
+    /// - On Darwin, the `ioctl()` request operates on a socket.
+    /// - On Linux, the `ioctl()` request operates on the `/dev/vsock` device.
+    ///
+    /// - Note: On Linux, ``local`` may be a better choice.
+    static func getLocalContextID(_ socketFD: NIOBSDSocket.Handle) throws -> Self {
+#if canImport(Darwin)
+        let request = CNIODarwin_IOCTL_VM_SOCKETS_GET_LOCAL_CID
+        let fd = socketFD
+#elseif os(Linux)
+        let request = CNIOLinux_IOCTL_VM_SOCKETS_GET_LOCAL_CID
+        let fd = open("/dev/vsock", O_RDONLY)
+        precondition(fd >= 0, "couldn't open /dev/vsock (\(errno))")
+        defer { close(fd) }
+#endif
+        var cid = Self.any.rawValue
+        try Posix.ioctl(fd: fd, request: request, ptr: &cid)
+        precondition(cid != Self.any.rawValue)
+        return Self(rawValue: cid)
+    }
+
+    /// Get the context ID of the local machine.
+    static func getLocalContextID() throws -> Self {
+        let socket = try Socket(protocolFamily: .vsock, type: .stream)
+        defer { try? socket.close() }
```

Comment:
`try!`, if open works, close must work or else we have a big issue (somebody closed our fd under our feat)

Reply:
Done—Good to know. Guess I need to tune into when it's OK to `try!`.